Add support for creating a FATE transaction by key

cshannon opened the pull request:
This updates FATE to support creating a transaction by a key which will be hashed to a transaction id. This can be used to guarantee only one transaction exists at the same time if the key is the same. Currently the key is just a byte array so anything can be used for the key. The key gets hashed to generate the transaction id (currently a long) and then encoded to be stored just like the existing randomly generated transaction ids.

I marked this as a draft PR for now to get some feedback as some changes are probably needed, in particular maybe changing the key from a byte array to something more strongly typed (see below) I added a few store tests but some tests are missing such as testing the collision detection. (Probably will require manually modifying some things in the store or ZK to simulate a collision or googling the algorithm to find two keys that cause a collision).

Something else is the hashing algorithm is using 128 bit murmur3 but converting to a positive long (since that's currently or tid) so the size is really only 63 bits and likely would lead to collisions at scale. Once the new FateId is added in #4191 we should explore changing the long tid into a byte array and using a 128 bit transaction id size. The algorithm could also be switched out to if desired (ie sha256)

For now the key is just a byte array to keep things simple but we can change it. The goal is to have unique fate operations a byte array is the most flexible as we can put anything into it. For example, if we wanted to create a key for a split operation for an extent we could serialize the operation name (ie split) and append it to the serialized KeyExtent. In the tests for now I'm only serializing the KeyExtent. We could create some utility methods to help with the serialization and deserialization.

However, we could also make the API more strongly typed and instead of a byte array as the key would create a new interface or class (maybe call it `FateKey`) and that could then contain an operation name and a KeyExtent and then handle the serialization. This could be nice as it would match up with the new `FateId` being created in #4191 but the downside is if we want different key types in the future it's not as flexible and we'd need to change or extend the API. If we just use an interface and had different key implementations it might be ok as we could add new key types in the future without changing the API.

This closes #4183

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java:
```diff
@@ -91,6 +94,27 @@ public long create() {
+          NodeExistsPolicy.FAIL);
+      // The key was already used to generate the tid but we still need to store it
+      // separate to check later for collision detection
+      zk.putPersistentData(getTXPath(tid) + "/" + TxInfo.TX_KEY, serializeTxInfo(key),
```
This comment is not really for changes in this PR, just thinking ahead.  Eventually we will want to store the key in the ZK node at `getTXPath(tid)`.  This is because can only atomically modify a single node in ZK.  So eventually we can atomically set the key in ZK and Accumulo using conditional mutations in both.

In #3964 I did something like this by creating a [new struct](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java#L89) that is the value of the ZK node at `getTXPath(tid)` and then Accumulo's util functions for atomically modifying can be used like was done [here](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/fate/ZooStore.java#L227).

So I think for this key and #4131 we will need a new struct in that ZK node.

Changes could be made in this PR for this, but don't have to be made.

cshannon replied:
This is now atomic in the latest updats

keith-turner commented:
> However, we could also make the API more strongly typed and instead of a byte array as the key would create a new interface or class (maybe call it FateKey) and that could then contain an operation name and a KeyExtent and then handle the serialization. 

There will only be two use cases (split and compaction commit) for the fate key at the moment.  So could have a class with static methods that support creating a FateKey object for the two uses case.

cshannon commented:
> There will only be two use cases (split and compaction commit) for the fate key at the moment. So could have a class with static methods that support creating a FateKey object for the two uses case.

That sounds good, we can expand it if we end up with more use cases in the future. I will work on re-working the byte array into a FateKey object this week as I think that is cleaner.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java:
```diff
@@ -216,7 +227,104 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+    } finally {
+      txStore1.delete();
+      txStore2.delete();
+    }
```
We need to try calling create after delete with the same key and ensure that works.

keith-turner replied:
May be better to do this in the in progress test because create when in new status will still succeed.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java:
```diff
@@ -216,7 +227,104 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+      txStore.delete();
+    }
+  }
+
```
Would be good to test the case of having a collision of the id when the keys differ.  Thinking we could do the following in test code so that no test specific modifications are needed to real code.

 1. Compute H1 = hash(K1) for key K1.  Use the same hash algo as store impls.
 2. In test code manually insert K2 where K1!=K2 in the data layer of the store using id H1 from above.
 3. Try to insert K1 using the store, should fail because K2 exist w/ same id.

Not sure about code organization for step 2 because it will need to be different test code depending on if Zookeeper or Accumulo store is being tested.

cshannon replied:
In order to make this easier to test I updated the store to allow providing a different hash code or FateId generated from the key and i replace it for the test to see if the collision detection worked

cshannon commented:
@keith-turner - I think this is ready for another review. Since you last reviewed it I added the FateKey type instead of a byte array, I made FateKey create method atomic for ZK and added a conditional mutation for Accumulo and I updated the tests (including testing the collision detection

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java:
```diff
@@ -217,10 +228,144 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+
+      // We have an existing transaction with the same key in progress
+      // so should not be allowed
+      assertThrows(IllegalStateException.class, () -> store.create(fateKey));
```
Could check the status is still IN_PROGRESS after the failed create

keith-turner commented on core/src/main/java/org/apache/accumulo/core/logging/FateLogger.java:
```diff
@@ -141,6 +142,15 @@ public int getDeferredCount() {
+      public FateId create(FateKey fateKey) {
+        FateId fateId = store.create(fateKey);
+        if (storeLog.isTraceEnabled()) {
+          storeLog.trace("{} created fate transaction", fateId);
```
```suggestion
          storeLog.trace("{} created fate transaction using key : {}", fateId, fateKey);
```

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/accumulo/AccumuloStore.java:
```diff
@@ -112,6 +121,16 @@ public FateId getFateId() {
+    var status = newMutator(fateId).requireStatus().putStatus(TStatus.NEW).putKey(fateKey)
+        .putCreateTime(System.currentTimeMillis()).tryMutate();
+
+    // TODO: Any reason to retry here?
```
An unknown status means that a failure happened w/ the RPC for the conditional mutation and its not know if it went through.  So need to scan in that case and see if it was created or not and possibly retry.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -239,12 +254,46 @@ public int getDeferredCount() {
 
+  @Override
+  public FateId create(FateKey fateKey) {
+    FateId fateId = fateIdGenerator.fromTypeAndKey(getInstanceType(), fateKey);
```
This code has a race condition with reading the status and then calling create.  Something could run between and cause create() to throw an exception.  Thinking it may be better to call create first and if it fails then to check into what the status/key currently is.

```java
FateId fateId = fateIdGenerator.fromTypeAndKey(getInstanceType(), fateKey);
try{
   create(fateId, fateKey);
}catch(ISE e){
   Pair<TStatus,Optional<FateKey>> statusAndKey = getStatusAndKey(fateId);
   TStatus status = statusAndKey.getFirst();
   if(status == NEW){
              Preconditions.checkState(tFateKey.isPresent(), "Tx key column is missing");
      Preconditions.checkState(fateKey.equals(tFateKey.orElseThrow()),
          "Collision detected for tid %s", fateId.getTid());
   } else {
      throw new ISE();
  }
}


```

cshannon replied:
Very good catch, I plan to refactor the create method into `Optional<FateTxStore<T>> createAndReserve(FateKey fateKey)` as suggested in this [comment](https://github.com/apache/accumulo/pull/4204#issuecomment-1934388313) and can also make these changes as well.

cshannon replied:
This was changed in the lasted update to fix the race condition, createAndReserve() change will be next

keith-turner reviewed (commented):
These changes look good.  I am still reviewing but I am mostly done.  May also need some changes in Fate class to do something like the following.

 1. Attempt to seed a split operation with a repo
 2. If the split exists with the given key and is not NEW then log a debug or trace.  This is expected to happen when the system is busy and operations are not running and we keep seeding them
 3. If there is a collision log an error

Not exactly sure how this should be done, trying to figure out case 2 above where we want to avoid spamming the logs w/ warn/error for an expected case.   Also wondering what the new Fate method should be to use this new functionality in the store, maybe something like the following.

```java
class Fate {
   public FateId startAndSeedTransaction(FateKey key, Repo repo) {
   }
}
```

The fate changes could be a separate PR or done in this PR.

cshannon commented on test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java:
```diff
@@ -217,10 +228,144 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+
+    FateTxStore<TestEnv> txStore = store.reserve(fateId1);
+    try {
+      try {
+        store.create(fateKey2);
+        fail("Expected IllegalStateException due to hashing collision");
+      } catch (Exception e) {
+        assertInstanceOf(IllegalStateException.class, e);
+        assertEquals("Collision detected for tid 1000", e.getMessage());
+      }
```
Thanks for the suggestion, I was trying to use assertThrows() but I wanted to check the message and didn't see a way to do that with assertThrows(). I didn't realize that method also returns the exception so that makes it easy to check.

keith-turner replied:
>  I didn't realize that method also returns the exception so that makes it easy to check.

Yeah I didn't know that for a while either when I was using it.  It is a really nice design by the junit team.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -239,12 +254,46 @@ public int getDeferredCount() {
+      // to reserve/seed as long as the existing key is the same and not different as that would
+      // mean a collision
+    } else if (status == TStatus.NEW) {
+      Preconditions.checkState(tFateKey.isPresent(), "Tx key column is missing");
```
I don't think there is a test for a collision between a random fate id and hash derived fate id.  Im 99.9% sure this code works, but was wondering if creating a test for that is worthwhile for detecting regressions.

cshannon replied:
@keith-turner - I just added a test for collision with key generated fate id and random id

The test works by creating the transaction using the key and then the test deletes the key outside of fate (either out of accumulo table or replaces the ZK node with an empty key). After removing the key the transaction appears to the store just like a normal random transaction so the test then verifies the failure when calling create again and also ensures the transaction can still be reserved and used since it was essentially "converted" to a random transaction.

keith-turner commented:
Stubbed out possible interaction between fate and the fate store for keyed operations in #4178.  Can see relevant changes in [FateStore](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/core/src/main/java/org/apache/accumulo/core/fate/FateStore.java#L45-L66), [Fate](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/core/src/main/java/org/apache/accumulo/core/fate/Fate.java#L339-L354), and [SplitTask](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitTask.java#L46-L56)

cshannon commented:
> Stubbed out possible interaction between fate and the fate store for keyed operations in #4178. Can see relevant changes in [FateStore](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/core/src/main/java/org/apache/accumulo/core/fate/FateStore.java#L45-L66), [Fate](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/core/src/main/java/org/apache/accumulo/core/fate/Fate.java#L339-L354), and [SplitTask](https://github.com/apache/accumulo/blob/8d5046f71624d14625101707ac8779f07b1cde01/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitTask.java#L46-L56)

I'll take a look in detail tomorrow, I see you added `Optional<FateTxStore<T>> createAndReserve(FateKey fateKey` and that is something I will likely add to this PR when I update it. I think it makes sense to add it now vs a follow on

keith-turner commented:
What lead to the createAndReserve call was trying to deal with the following race condition when I was trying to sub out using this feature.

```java
class Fate {
  seed(..., FateKey key,...) {
       var fateId = create(key);
       // another thread could sneak in here
        var fateTxStore = store.reserve(fateId);
        try{
            // between create and reserve, its possible something happened with the key so now that we hold the reservation check that things are as expected
            store.checkKeyIsStillSame(fateId,key);
        }finally{
            fateTxStore.unreserve();
       }
   }

```

So hoping having the store create and reserve can avoid this.

cshannon commented:
I made a bunch of updates and addressed everything and changed create(fateKey) to createAndReserve(fateKey) (which technically is implemented by reserving first in memory and then creating). I also added a test for verifying that collisions when using a key and with random ids work as expected.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -295,16 +296,28 @@ public Optional<FateTxStore<T>> createAndReserve(FateKey fateKey) {
-        Preconditions.checkState(create(fateKey) != null,
-            "Unexpected null FateId when creating and reserving fateKey %s", fateKey);
+        if (create(fateKey).isPresent()) {
+          txStore = reservedTxStore;
```
Could modify the code to keep the return value of create and validate it.

```suggestion
           Precondition.checkState(fateId.equals(fateIdFromCreate));
          txStore = reservedTxStore;
```

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -239,12 +254,101 @@ public int getDeferredCount() {
+        }
+      }
+    } else {
+      // Could not reserve so return empty
```
Created the following after seeing another log stmt in this PR.  Thought it would be good to log something here like it does elsewhere.  These trace stmts will be nice for tracking down bugs

```suggestion
      // Could not reserve so return empty
      log.trace("Another thread currently has transaction {} key {} reserved", fateId,
            fateKey);
```

keith-turner commented on core/src/test/java/org/apache/accumulo/core/fate/TestStore.java:
```diff
@@ -32,25 +32,54 @@
 
+  @Override
+  public Optional<FateTxStore<String>> createAndReserve(FateKey key) {
+    HashCode hashCode = Hashing.murmur3_128().hashBytes(key.getSerialized());
```
This TestStore is only used by unit test. If the existing unit test does not call thsi method then could throw an UnsupportedOpException.
